Collect values identifier in Ir.value

n-osborne opened the pull request:
Building on the new intermediary representation and using the reserr monad of PR #5, we collect the identifiers of the values that satisfy three criteria:
1. they are not constant (there is nothing to be tested by `qcheck-stm` here)
2. they don't return a value of the same type as the system under test (`qcheck-stm` does not support it yet)
3. they don't have multiple argument of the same type of the system under test (`qcheck-stm` does not support it yet)

These criteria are meant to be completed when we fill the Ir.value record with more information.

n-osborne committed:
Add a pretty printer and the identifier argument to Ir.value

n-osborne committed:
Get the identifier of pertinent values

Begin to fill the Ir.value record with information from the Gospel typed
ast. For now, pertinent values are:
- values that are not constant (can't test anything)
- values that does not return a value of the type of the system under
  test (qcheck-stm does not support it yet)
- values that does not have multiple arguments of the type of the system
  under test (qcheck-stm does not support is yet)

These criteria are meant to be completed later when we fill the value
record with other information.

n-osborne committed:
Test the filtering of values

Complete the tested lib and print:
- all the errors when there is
- the name of the value when it passes the filtering.

shym commented on plugins/qcheck-stm/src/ir_of_gospel.ml:
```diff
@@ -0,0 +1,58 @@
+open Reserr
+open Config
+
+let lb_arg_is_of_type ts lb =
```
It might be interesting there, at least as a TODO, to ensure that the arguments are compatible with our expectancy. When testing `array` with `char` as placeholder for all `'a` types, then a function could be tested if it has an argument of type `'a array` or `char array` but not `int array`.

n-osborne replied:
I'll put is here as a TODO, as I expect the way we store the type information about the system under test in the `Config.t` to change in the near future.

shym commented on plugins/qcheck-stm/src/reserr.ml:
```diff
@@ -0,0 +1,31 @@
+  | Value_have_no_sut_argument of (Ppxlib.location * Ident.t)
+  | Value_have_multiple_sut_arguments of (Ppxlib.location * Ident.t)
+
+let error_to_string = function
```
We’ll eventually use the location info, I expect?
I’d like us to merge this back into `Warnings` at some point (not in this PR).

n-osborne replied:
Yes, this version is just enough for testing purposes. But it will be merged in `Warnings` where there is all what is needed to use the location information.

n-osborne committed:
Add todo comments about argument type compatibility

When scanning a value's type, we should do the checks also looking at
the argument type:

if the system under test has type `char t`, then `char t` or `'a t` are
  good, but not `int t`.
